### PR TITLE
DeepSeek V4: profit-gated adaptive K3 verifier width

### PR DESCRIPTION
**Pull request by davidtai**
## What changed

Adds the fixed DeepSeek-V4 max-K3 adaptive verification policy. The installed lane prebinds exact M2/M3/M4 verifier routes and contracts after low draft margins while retaining the existing target authority and `capture_commit`/stock/committed-history contract. The benchmark wrapper is portable: deployment locations are explicit operator configuration, while artifact, topology, route, and MLX identity checks remain fail-closed.

## Why

On the canonical single-load, primed 328-token → 256-generated K3 bracket this is a relative workload winner: 38.0405936601 tok/s versus a 36.7252878598 tok/s two-control mean (+1.3153058003 tok/s, +3.58%). It does **not** reach the 40 tok/s target; it remains 1.9594 tok/s short.

## How to verify

```sh
uv run --extra dev pytest -q tests/test_deepseek_v4_*.py tests/test_runtime_deepseek_v4_o_lora.py tests/test_generation_sustained.py tests/test_context_copy_stats.py
uv run --extra dev ruff check mtplx/deepseek_v4_adaptive_width.py mtplx/generation.py scripts/deepseek_v4_adaptive_width_guarded.py scripts/deepseek_v4_mtpk_bench.py tests/test_deepseek_v4_adaptive_width_bracket.py tests/test_deepseek_v4_adaptive_width_policy.py
```

For the production bracket, use `scripts/deepseek_v4_adaptive_width_guarded.py` inside the existing exclusive-GPU/service restoration workflow and provide `MTPLX_DSV4_PYTHON`, `MTPLX_DSV4_GUARDED_RUNNER`, `MTPLX_DSV4_QUALITY_PLIST`, `MTPLX_DSV4_BENCH_DIR`, `MTPLX_DSV4_MODEL_PATH`, and `MTPLX_DSV4_PROMPT_FILE`.

## Notable decisions

- Stacked on #223 (O-LoRA gather); #216 is already merged.
- The source intentionally has no developer-machine absolute paths; model artifact/config identity, not path spelling, is the canonical guard.
- The full receipt reports an allowed BF16 M3-width near-tie at generated index 6 (603 → 305) and its propagated tail. HumanEval is deferred by explicit policy; this PR does not claim quality equivalence.
- The detailed benchmark receipt and restoration evidence are in the PR comment.

### CODE REVIEW

**Comment by davidtai** [comment]
## Canonical K3 benchmark receipt (full workload; relative win, not 40 tok/s)

This is the verified full 328-token prompt → 256 generated-token workload, not a microbenchmark or profiler replay. It was run once in one process (PID `50094`) with one model load/object (`4613715632`), one discarded complete K3 primer, then `K3-C0 → ADAPTIVE-B → K3-C1`. Every measured arm used greedy stock target verification, `capture_commit`, committed-history policy, Stage-4 fused attention, HC compile, Sinkhorn, O-LoRA gather, MoE-tail, BF16 activation storage, canonical model/MTP bank, and official MLX `0.31.2`.

Exact recorded child command:

```sh
python /Users/davidtai/projects/OpenSourceWTF/mtplx-hy3-ssd/.worktrees/deepseek-v4-adaptive-width-policy/scripts/deepseek_v4_mtpk_bench.py --adaptive-width-bracket --model /Users/davidtai/models/DeepSeek-V4-Flash-2bit-DQ-mtp --prompt-file /Users/davidtai/projects/OpenSourceWTF/bench/deepseek-v4/smoke-2bitdq-20260731-prompt2.txt --max-tokens 256 --depths 3 --verify-strategy capture_commit --verify-core stock --mtp-history-policy committed --warmup-tokens 0 --out /Users/davidtai/projects/OpenSourceWTF/bench/deepseek-v4/adaptive-width-policy-20260802T203123Z-61a4d9ca
```

Workload identity: prompt SHA-256 `ee94397faa812c91d5f1a0ee17c5bb6ca6032883653591dd33d4cfddb737ac33`, 328 input tokens, 256 output tokens, source `61a4d9cac67d5feed9143a9efcfc1e3024b136d7`, production decode manifest `8e8bbadee0a25ff13736f475ef5882d03b659dc9be1a04dbfbbfb6d854af714b`.

| arm | tok/s |
|---|---:|
| K3-C0 | 36.2696765208 |
| adaptive B | **38.0405936601** |
| K3-C1 | 37.1808991988 |

Two-control mean: `36.7252878598` tok/s; control drift: `0.911222678` tok/s; preregistered promotion floor: `37.6365105378` tok/s. B gains `+1.3153058003` tok/s (`+3.58%`) and clears its floor by `0.4040831223` tok/s. It remains **1.9594063399 tok/s below 40**; this PR makes no absolute-40 claim.

Engagement/counters for B: accepted/rejected `152/59`; drafted by depth `[87, 86, 65]`; 91 target verifies; 238 MTP forwards/drafted tokens; 87/87 eligible policy events plus four context copies; width histogram `K1/M2=1`, `K2/M3=21`, `K3/M4=65`; peak `94.454744216 GiB` (active-after-load `93.392828424 GiB`).

### Quality caveat, explicitly accepted for this lane

The only accepted source-bound cause is a BF16 M3-width near tie at generated index 6 / absolute position 334: control `603` → adaptive `305`; control/candidate target logits `29.375/29.25`, gaps `0.125/0`. The propagated tail begins at index 7 and has 246 divergent tokens of 249 compared (247/256 total). The A/B/C diagnostic showed B/M3-live equals C/M3-explicit-commit exactly, so explicit commit is not the cause. HumanEval is deferred by the agreed policy. This is therefore a measured relative performance winner with a documented precision caveat, **not** a claim of output-quality equivalence.

Supporting artifacts (all status 0):

- Primary: `bench/deepseek-v4/adaptive-width-policy-20260802T203123Z-61a4d9ca.json`, SHA-256 `9c09c108b21fe64319369035c2e4f2a1307222694f604fa6a65b298dd105f878`
- Transcript: `bench/deepseek-v4/adaptive-width-policy-20260802T203123Z-61a4d9ca.txt`, SHA-256 `7b45cab3cea7cc3450584034eb16688f015b3d3390f016883c5e3782e33efc03`
- Restoration postflight: `bench/deepseek-v4/adaptive-width-policy-20260802T203123Z-61a4d9ca-postflight.json`, SHA-256 `2edc8622bba9c0826972589b1abb7b7ab4ac915d3fd4b322fd0bb2324d609cc3`
- Near-tie diagnostic primary/postflight: `bench/deepseek-v4/adaptive-neartie-probe-20260802T200528Z-8a83f1e2.json` / `...-postflight.json`, SHA-256 `8d537538453c649eda177e6a2017a47bcdc3ffb830de13db34728c36e00e6bf9` / `d49b7c2c1a9a3e08e6426cfa9ea664702ea1af2a9be4616f526cddea7f538e78`

The guarded postflight acquired/held/released the exclusive GPU lock (inode `208479940`), kept `iogpu.wired_limit_mb=114688`, and restored the exact Qwen Quality singleton; `/v1/models` and a real READY chat ended with `finish_reason=stop`. No knob was raised.

This stack depends on #223; #216 is already merged. The new portable wrapper deliberately requires its deployment paths via explicit `MTPLX_DSV4_*` environment variables instead of publishing a developer-specific filesystem contract.

**Comment by davidtai** [comment]
Superseded by the consolidated DeepSeek V4 shape-specific K3 optimization PR #223. This PR body and its benchmark evidence have been archived verbatim in #223.
